<center>
<h1>Chapter 3</h1>
</center>

<hr>

<center><h1>Roadmap</h1></center>

<img src='https://www.dropbox.com/s/38llhv7ckhrlvuj/Screenshot%202020-04-12%2010.24.39.png?raw=1'>

#1. I built a library just for you
<img src='https://www.dropbox.com/s/9fcc1crlxp19ijt/major_section.png?raw=1' width='300'>

Before looking at predicting survival of Titanic passengers, we will need a couple of new functions. I wrote these functions for you. So all you have to do is figure out how to use them!

First, load the library that I built. 
I called my library uo_puddles :)

<img src='https://www.dropbox.com/s/zv84ew76m7ptyrh/puddles.jpeg?raw=1' height=100>


In [0]:
#flush the old directory - if you get a "cannot remove" error that is ok - ignore it
!rm -r  'uo_puddles'

In [0]:
my_github_name = 'uo-puddles'  #this is one of my github accounts
my_library_name = 'uo_puddles' #this is name of repository I keep on github - easy to set up!

In [0]:
clone_url = f'https://github.com/{my_github_name}/{my_library_name}.git'  #create the url to get the library
clone_url

'https://github.com/uo-puddles/uo_puddles.git'

In [0]:
#this adds the library to colab operating system - it is another CLI command
!git clone $clone_url  


Cloning into 'uo_puddles'...
remote: Enumerating objects: 198, done.
remote: Counting objects: 100% (198/198), done.
remote: Compressing objects: 100% (162/162), done.
remote: Total 198 (delta 115), reused 64 (delta 33), pack-reused 0
Receiving objects: 100% (198/198), 49.68 KiB | 275.00 KiB/s, done.
Resolving deltas: 100% (115/115), done.


In [0]:
import uo_puddles.uo_puddles as up  #now that operating system has it, you can import it.

###Explanation of what you just did

In the first code cell you see me doing this:
<pre>
!rm -r  'uo_puddles
 </pre>
  What does that explanation point ("bang" in jargon) do? It is kind of interesting. It tells jupyter we want to talk to the underlying operating system, Linux. When we talk to the operating system in this way, we are using something called the "command line interface" or CLI. Normally you would use the CLI to talk to your laptop operating system. In this case, we are talking to the colab operating system. Why? Because the operating system stores the libraries we want. And I am saying remove `up_puddles`. Why? Because I want to make sure I have the freshest version. You will see that a little later I  restore `uo_puddles` directly from my github account. This has the freshest version. Once the operating system has it, I can import it as normal.

**Important note**: you can do the same. You can store a library on your own github account. Give it a name. Then mostly duplicate my steps below to import it. Let me know if you are interested in doing this and I can help you out. It is pretty dang cool to have your own library. You can even build it for your research group and share it with them.

#2. First goal: compute how similar two passengers are
<img src='https://www.dropbox.com/s/9fcc1crlxp19ijt/major_section.png?raw=1' width='300'>


You saved your work from last chapter in Google Sheets. And got a url. Now you can use that url below.

In case you were unable to generate your own url, you can use mine: https://docs.google.com/spreadsheets/d/e/2PACX-1vShZZ4EVYQQKWcdQ1JzXxIFj8PER0zk9LwF67YY1BR3iXkWBopwFrGS8-VL1bSACKCRBgOsnY3rAqHU/pub?output=csv

In [0]:
import pandas as pd

In [0]:
#your url from last chapter
url = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vShZZ4EVYQQKWcdQ1JzXxIFj8PER0zk9LwF67YY1BR3iXkWBopwFrGS8-VL1bSACKCRBgOsnY3rAqHU/pub?output=csv'


In [0]:
letters_table = pd.read_csv(url)



In [0]:
letters_table.head()

Survived                                               Name  Length  ...  x  y  z
0         0                            Braund, Mr. Owen Harris      23  ...  0  0  0
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...      51  ...  0  2  0
2         1                             Heikkinen, Miss. Laina      22  ...  0  0  0
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)      44  ...  0  2  0
4         0                           Allen, Mr. William Henry      24  ...  0  1  0

[5 rows x 29 columns]

##General idea: crowdsourcing

<img src='https://i1.wp.com/www.feedough.com/wp-content/uploads/2017/11/CROWDSOURCING-01-1.png' height='150'>

What we want to do is to treat some of the passengers of the Titanic as the "crowd". When we see a new passenger, we ask which of the passengers in the crowd is most similar to the new passenger. If we can do this, we are most of the way to our final goal of predicting survival.

If you were standing on the gangplank looking at passengers boarding, you would probably look at gender and age and "social class". 

<center>
<img src='http://images2.fanpop.com/image/quiz/322000/322400_1259414563406_450_300.jpg' height=150>
</center>

But we threw those all away! What do we have left to work with? The length of their names and the counts of the 26 letters. So we will plow ahead with that.

At the most basic, we need to compute a similarity measure between 2 passengers.


Let's say I wanted to compare Mr. Braund to Mr. Allen. One way to do that is to look at the list of numbers following their names. There are 27 numbers. What if we put them into a list. So we would have 2 lists of numbers to compare.

<img src='https://www.dropbox.com/s/tlhcphro1ntymrh/Screenshot%202020-01-31%2013.10.14.png?raw=1'>

Let's do that, create the 2 lists of numbers. I can get a row from the table with `.iloc[0].tolist()`. You can replace the 0 with any legal row index to get that row.

A word about the `.iloc[0]` piece. It is the row indexer for a table. We saw the column indexer in chapter 2. This is the row indexer for a table. So in same conceptual ballpark, but now working with a 2D data type.

It would be nice if we could do this instead, but no can do. You will get an error:
<pre>
letters_table[0]  #instead use .iloc[0] for row indexer
</pre>

In [0]:
braund_row = letters_table.iloc[0].tolist()
braund_number_list = braund_row[2:]
print(braund_row)  #[0, 'Braund, Mr. Owen Harris', 23, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 2, 1, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]

[0, 'Braund, Mr. Owen Harris', 23, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 2, 1, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]


The first item is the value of Survived and second item is the Name. We want everything past these 2. We can slice to get that.

In [0]:
braund_number_list = braund_row[2:]  #3rd item to the end
print(braund_number_list)

[23, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 2, 1, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]


Repeat for Mr. Allen.

In [0]:
allen_row = letters_table.iloc[4].tolist()
print(allen_row)

[0, 'Allen, Mr. William Henry', 24, 2, 0, 0, 0, 2, 0, 0, 1, 2, 0, 0, 4, 2, 2, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, 1, 0]


We can slice to get just the numbers.

In [0]:
allen_number_list = allen_row[2:]  #3rd item to the end
print(allen_number_list)

[24, 2, 0, 0, 0, 2, 0, 0, 1, 2, 0, 0, 4, 2, 2, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, 1, 0]


#3. First insight: recast as a Geometry problem
<img src='https://www.dropbox.com/s/9fcc1crlxp19ijt/major_section.png?raw=1' width='300'>

Our goal is to compute how "similar" the 2 lists of numbers are.
Let's take the view that each list of numbers is actually a point in 27-dimension space. Sounds kind of scary already! Let me give you the intuition pretending we only have 2 numbers in our list. Call the two lists A and B. We can view a list as a point on a 2D plot as shown below. Ignore the angle theta for now.

<img src='https://www.dropbox.com/s/7rtuzw37hgl1oi8/ed_vs_cos.png?raw=1' height=300>

We can use the distance d as a measure of their similarity. The smaller the distance, the closer they are. If the distance is 0, they are the same point.

###Good news!
<img src='https://www.sapaviva.com/wp-content/uploads/2017/06/6S.-Euclid-of-Alexandria-ca.320-275-BC-225x225.jpg' height='100'>

Euclid (circa 300BC) figured it out for us. He came up with a formula for computing d. It defines the "Euclidean distance" between 2 points (here called p and q instead of A and B) as follows.

  <p>
<img src='https://www.dropbox.com/s/9wao0kf3u32i3e9/euclidian.png?raw=1'>

Linking our plot above into this:

* n = 2. We have 2 values in each list.
* q1 is the same as x1, p1 is the same as x2.
* q2 is the same as y1, p2 is the same as y2.

The capital (Greek) letter sigma represents summation. I'll write the code out long hand for you.

BTW: this is a useful site that translates mathematical notation into Python: https://amitness.com/2019/08/math-for-programmers/.


In [0]:
p = [1,3]  #just guessing from the plot
q = [3,2]  #ditto
n = 2

In [0]:
greek_sigma = (p[0]-q[0])**2 + (p[1]-q[1])**2  #Python uses ** for exponentiation. In this case, squaring a quantity.
greek_sigma

5

In [0]:
d = greek_sigma**.5  #dredging up some old algebra to take the square root using exponentation
d

2.23606797749979

###Does 2.23606797749979 mean the 2 points are close?

Euclidean distance just gives you the distance. Makes no value judgement. We do know that a distance of 0 says the points are exactly the same. So that is the min value. But there is no limit on the upper value.

#Stop!
<img src='https://www.dropbox.com/s/kfnmns0f60q8b8z/red_line.png?raw=1' width='300'>

I am going to cut you loose on Assignment 1 today.

On Wednesday I will go over a bit more material and ask you to start on the remaining assignments.

All assignments turned into Canvas for this chapter.

#Assignment 1 (all assignments turned in this chapter)
<img src='https://www.dropbox.com/s/3uyvp722kp5to2r/assignment.png?raw=1' width='300'>

I would like you to use a loop to calculate the euclidean distance between braund and allen. Some hints:

* You might want to start by bringing in the "create new list from old list" gist.

* I can see some similairty with this problem and the problem of chapter 2 of taking the average.

* You can let taking the square-root be the last step in the process.

In [0]:
#your code here
n = len(braund_number_list)
square_diff = []

for i in range(n):
  p = braund_number_list[i]
  q = allen_number_list[i]
  square_diff.append((p-q)**2)

d = sum(square_diff)**.5
print(d)

5.477225575051661


Check your answer against mine.

In [0]:
d  #5.477225575051661

5.477225575051661



#End Assignment 1

##Getting ready for Assignment 2

Among all the passengers, who is the closest to Mr. Braund? For closest we will use Euclidean distance. 
But first, we need a new concept. Check this code out.

In [0]:
test_point = [1,2,3]  #so n = 3 because each point has 3 values, i.e., dimension 3.
other_points = [[2,3,4], [4,5,6], [7,8,9]]  #3 other points

###Wait a minute!

You know about lists of numbers and list of strings. What's up with this?
<pre>
[[2,3,4], [4,5,6], [7,8,9]]
</pre>
Yup, it is a list of lists. If you turn it on its side, it looks more like a table, right?
<pre>
[[2,3,4],
 [4,5,6],
 [7,8,9]]
</pre>
Kind of similar to this view but on a smaller scale.

<img src='https://www.dropbox.com/s/1oznqiacdqz25dy/Screenshot%202020-01-31%2016.14.14.png?raw=1'>

#Let's sum a column for practice

Let's sum column 1 (actually the 2nd column) in other_points.


In [0]:
print(other_points)

[[2, 3, 4], [4, 5, 6], [7, 8, 9]]


In [0]:
column = 1

##Bring in loop pattern

Pattern to use is new list from old list.

In [0]:
#old list is other_points  [[2, 3, 4], [4, 5, 6], [7, 8, 9]]

old_list = other_points
n = len(old_list)

new_list = []
for i in range(n):
  item = old_list[i]
  new_item = item[column]
  new_list.append(new_item)

print('done')
print(old_list)
print(new_list)
sum(new_list)

done
[[2, 3, 4], [4, 5, 6], [7, 8, 9]]
[3, 5, 8]


16

##How about building a matrix

In [0]:
old_list = [1,2,3]

We will get to a point where we need to reshape a 1D list like old_list into a 2D list (i.e., a nested list) with 1 column and 3 rows. This is another form of data wrangling: getting our data in right shape to feed to a machine-learning algorithm.

Here is our target:
<pre>
Old: [1,2,3]
New: [[1], [2], [3]]
</pre>
Or stacked vertically:
<pre>
[[1],
 [2],
 [3]]
 </pre>

Let's see if we can do it using new list from old list pattern.

In [0]:
# matrix = list of lists (python does not recognize / have variable class for matrices)
# old_list = [1,2,3]
n = len(old_list)

new_list = []
for i in range(n):
  item = old_list[i]
  new_item = [item]
  new_list.append(new_item)

print(old_list)
print(new_list)

[1, 2, 3]
[[1], [2], [3]]


#Assignment 2
<img src='https://www.dropbox.com/s/3uyvp722kp5to2r/assignment.png?raw=1' width='300'>

I'd like you to produce what I have in red in the image above except for the whole table, not just first 5 rows. So you will have a list of lists just like in `other_points`. But now each inner-list is holding 27 values instead of 3. Once we have this, we can move on to the next step.

I'm going to ask you to call your list of lists `all_passengers_matrix`. I will start using the term "matrix" to designate a list of lists. The term has no special meeting in Python; I'm using it more in its abstract mathematical sense. It's fine to just think of list of lists when I say matrix.

Once you are finished, you should get this:

<pre>
for i in range(5):
  print(all_passengers_matrix[i])

[23, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 2, 1, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]
[51, 2, 2, 2, 1, 4, 1, 3, 2, 2, 1, 0, 2, 2, 3, 2, 0, 0, 5, 3, 1, 1, 0, 0, 0, 2, 0]
[22, 2, 0, 0, 0, 2, 0, 0, 1, 4, 0, 2, 1, 1, 3, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0]
[44, 3, 0, 1, 0, 6, 1, 0, 2, 1, 1, 0, 5, 2, 0, 0, 1, 1, 2, 2, 2, 2, 0, 0, 0, 2, 0]
[24, 2, 0, 0, 0, 2, 0, 0, 1, 2, 0, 0, 4, 2, 2, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, 1, 0]
</pre>
Looks just like the image - nice.

At this point I brought in my "create new list from old list" gist. The old list is actually the `letters_table`. Here are a few things that you might need.

* To convert a row `i` in the table to a list of numbers, use: `letters_table.iloc[i].tolist()`. What is the value of `i`? Whatever you want it to be, as long as it is a legal row in the table. We have already used this to get the rows for Braund and Allen earlier, right?

* Each row actually has 29 values. You need to skip the first 2. Look back at Braund-Allen example to see how to accomplish this.

In [0]:
braund_row = letters_table.iloc[0].tolist()
braund_number_list = braund_row[2:]

In [0]:
letters_table.head()

Survived                                               Name  Length  ...  x  y  z
0         0                            Braund, Mr. Owen Harris      23  ...  0  0  0
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...      51  ...  0  2  0
2         1                             Heikkinen, Miss. Laina      22  ...  0  0  0
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)      44  ...  0  2  0
4         0                           Allen, Mr. William Henry      24  ...  0  1  0

[5 rows x 29 columns]

In [0]:
#your code

# old_list = letters_table
n = len(letters_table)

all_passengers_matrix = []
for i in range(n):
  passenger = letters_table.iloc[i].tolist()
  passenger_num_list = passenger[2:]
  all_passengers_matrix.append(passenger_num_list)


In [0]:
#your output should match mine above

for i in range(5):
  print(all_passengers_matrix[i])

[23, 2, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 2, 1, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]
[51, 2, 2, 2, 1, 4, 1, 3, 2, 2, 1, 0, 2, 2, 3, 2, 0, 0, 5, 3, 1, 1, 0, 0, 0, 2, 0]
[22, 2, 0, 0, 0, 2, 0, 0, 1, 4, 0, 2, 1, 1, 3, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0]
[44, 3, 0, 1, 0, 6, 1, 0, 2, 1, 1, 0, 5, 2, 0, 0, 1, 1, 2, 2, 2, 2, 0, 0, 0, 2, 0]
[24, 2, 0, 0, 0, 2, 0, 0, 1, 2, 0, 0, 4, 2, 2, 0, 0, 0, 2, 0, 0, 0, 0, 1, 0, 1, 0]


#Assignment 3
<img src='https://www.dropbox.com/s/3uyvp722kp5to2r/assignment.png?raw=1' width='300'>

Create a list of differences between Mr. Braund and all 891 passengers using Euclidean distance as the measure. 

Retain both the distance and the row number that goes with that distance.



I'm going to give you a little help. There is a euclidean distance function already built in uo_puddles. Check it out.

In [0]:
up.euclidean_distance(braund_number_list, allen_number_list)  #5.477225575051661

5.477225575051661

##Here is what I want

I want you to create a list called `all_distances_matrix`. It will hold all the Euclidean distances from `braund_number_list` to the rows (also number lists) in `all_passengers_matrix`. Each individual list in all_distances_matrix will be a list of 2 items: the distance and the row index. Here is what I came up with and it should be your target.
<pre>
all_distances_matrix[:5]

[[0.0, 0],                 #compare Braund to Braund
 [28.792360097775937, 1],  #compare Braund to Cumings (row 1 in table)
 [6.244997998398398, 2],   #compare Braund to row 2 in table
 [22.825424421026653, 3],  #etc.
 [5.477225575051661, 4]]
 </pre>
 You can read this as the distance between `braund_number_list` and the first passenger number list is 0. That is because the first passenger is Braund!

The distance between braund_number_list and the second passenger number list is 28.792360097775937. Etc.

###I used my gist

I'm trying to create a new list `all_distances_matrix` from an old list `all_passengers_matrix`. A good fit for my gist. The one twist is I am appending a new 2-item list (i.e., distance and row number) instead of just a single item.

In [0]:
#Here is our old list
print(braund_number_list)

In [0]:
#your code here
# braund_number_list
# n = len(all_passenger_matrix)

all_distances_matrix = []
for i in range(n):
  passenger_num_list = all_passengers_matrix[i]
  euc_dist = up.euclidean_distance(braund_number_list, passenger_num_list)
  all_distances_matrix.append([euc_dist,i])


In [0]:
all_distances_matrix[:5]  #should match against mine above

[[0.0, 0],
 [28.792360097775937, 1],
 [6.244997998398398, 2],
 [22.825424421026653, 3],
 [5.477225575051661, 4]]


#End Assignment 3

#Once you finish you are most of the way there

Here is what I am going to do. I am going to sort all_distances_matrix on distance. This will be extremely useful to us for looking at who is closest to Braund.

Python gives us a function for doing sorting called `sorted`. And if we don't tell it otherwise, it will use the 1st item of each pair to sort on. And it will sort in ascending order. Exactly what we want. BTW: sorted returns a copy of the list. So original list is unchanged.

In [0]:
sorted_distances_matrix = sorted(all_distances_matrix)  #sorted will default to using the first item in our 2-item lists

In [0]:
'''
[[0.0, 0],
 [3.1622776601683795, 477],
 [3.1622776601683795, 482],
 [3.1622776601683795, 804],
 [3.1622776601683795, 806],
 [3.4641016151377544, 75],
 [3.4641016151377544, 93],
 [3.4641016151377544, 508],
 [3.4641016151377544, 698],
 [3.4641016151377544, 766]]
 '''

sorted_distances_matrix[:10]

[[0.0, 0],
 [3.1622776601683795, 477],
 [3.1622776601683795, 482],
 [3.1622776601683795, 804],
 [3.1622776601683795, 806],
 [3.4641016151377544, 75],
 [3.4641016151377544, 93],
 [3.4641016151377544, 508],
 [3.4641016151377544, 698],
 [3.4641016151377544, 766]]

##Who is passenger 477?

The passenger in row 477 bubbled to the top. He or she is actually tied with 3 others as having the smallest distance to Braund.

Do you remember how to pull a row out of the table? Might be good to have in a gist.

In [0]:
p477 = letters_table.iloc[477].tolist()  #go back to original table and see
print(p477)

[0, 'Braund, Mr. Lewis Richard', 25, 2, 1, 1, 2, 1, 0, 0, 1, 2, 0, 0, 1, 1, 1, 0, 0, 0, 4, 1, 0, 1, 0, 1, 0, 0, 0]


###Hmmmm

Maybe his brother? He also perished.

##Take a look at the top 4 matches in terms of survival

In [0]:
top4 = [477, 482, 804, 806]

for i in range(4):
  index = top4[i]  #477, etc.
  row = letters_table.iloc[index].tolist()
  outcome = row[0]  #first value in row is Survived value, 0 or 1
  print(outcome)

0
0
1
0


##Kind of interesting

Three out of four of the top matches perished. If you view them as important because they were a close match, then maybe you can view them as voting whether Mr. Braund survived or perished. We are well on our way to our first machine learning model!

##Let's take stock.

* We are focusing on a specific passenger, Braund (passenger 0).

* We want to know what other passengers are similar to him.

*  We wrangled passenger names to get 27 new numeric columns.

* Once we have number lists, we can employ Euclidean distance to help with our similarity questions.

* We looped through the number lists of all passengers (including Braund) and computed the Euclidean distance from Braund.

* We sorted the list ordered on smallest distances.

* We checked out the closest and saw it was his brother (maybe)!

The sorted list could be interesting on its own. Knowing when one piece of text (e.g., a document) is similar to another piece of text (another document) seems useful. Maybe we are trying to find plagarized political speeches. We could build our 27 element list for each speech. Then follow our steps to get ordered distances.

But we have other fish to fry. Our end goal is to predict survival of passengers. And the sorted list gives us a huge headstart on our first machine-learning algorithm called K Nearest Neighbor or simply KNN. See you in next chapter.